### ISSUE TITLE
Lua custom writers cannot create or edit template variables

### ISSUE THREAD

**Issue by fasiha** [jgm/pandoc#2358]
Creating an issue for the root cause of #2353 to track progress: unlike Haskell writers, Lua custom writers cannot create or edit template variables.

**Comment by fasiha** [comment]
Workaround: write a filter that creates or edits template variables.

**Comment by jgm** [comment]
To clarify this issue, the proposal is to provide a way to set template variables within a lua custom writer.

**Comment by crmackay** [comment]
This would be useful, yes. This way writer can set variables, based upon what it encounters when rendering the AST elements. eg: there is a block of code with a language attribute, ok set the flag that includes all of the highlighting macros.

```
$ pandoc -t customwriter.lua --template=myTemplate.latex my_file.md
```

Somewhat related: giving the custom writer access to the variables (and metadata) up front would be helpful too. For example when processing headings, there is no way currently to see if the top-level should be a chapter or a section. This could be inferred from the `documentclass` variable, but this variable is not available to the `Header()` function? I have wondered if it was possible to have a Metadata() or Variables() functions in the lua custom writers file, which are called first. They could then be used to set lua global variable that each AST-element function could then utilize.

Finally, in the same vein, being able to see what command-line options were used would also be useful. 

(I can split the last two things out into issues or pandoc-discuss discussions if that's easier)

**Comment by morganwillcock** [comment]
> Workaround: write a filter that creates or edits template variables.

Could I clarify, how would a Lua filter create a template variable? I can see how to add metadata but metadata variables aren't directly usable by the template (unless I've misunderstood).

**Comment by tarleb** [comment]
Fixed in pandoc 2.17 and later: template variables can now be set by returning them as a second result from `Doc`. See <https://pandoc.org/custom-writers.html#template-variables>.
